applications/ConstitutiveLawsApplication/custom_constitutive/masonry_orthotropic_damage_plane_stress_2d.cpp

```diff
@@ -0,0 +1,926 @@
+            return true;
+        if (rThisVariable == THRESHOLD_COMPRESSION)
+            return true;
+        return false;
+    }
+
+    /// GetValue variable <double>
+    double& MasonryOrthotropicDamagePlaneStress2DLaw::GetValue(
+        const Variable<double>& rThisVariable,
+        double& rValue)
+    {
+        rValue = 0.0;
+        // TENSION
+        if (rThisVariable == DAMAGE_TENSION)
+            rValue = mDamageTension;
+        else if (rThisVariable == UNIAXIAL_STRESS_TENSION)
+            rValue = mUniaxialStressTension;
+        else if (rThisVariable == THRESHOLD_TENSION)
+            rValue = mThresholdTension;
+        // COMPRESSION
+        else if (rThisVariable == DAMAGE_COMPRESSION)
+            rValue = mDamageCompression;
+        else if (rThisVariable == UNIAXIAL_STRESS_COMPRESSION)
+            rValue = mUniaxialStressCompression;
+        else if (rThisVariable == THRESHOLD_COMPRESSION)
+            rValue = mThresholdCompression;
+        return rValue;
+    }
+
+    /// SetValue variable <double>
+    void MasonryOrthotropicDamagePlaneStress2DLaw::SetValue(
+        const Variable<double>& rVariable,
+        const double& rValue,
+        const ProcessInfo& rCurrentProcessInfo)
+    {
+        // TENSION
+        if (rVariable == DAMAGE_TENSION)
+            mDamageTension = rValue;
+        else if (rVariable == UNIAXIAL_STRESS_TENSION)
+            mUniaxialStressTension = rValue;
+        else if (rVariable == THRESHOLD_TENSION)
+            mThresholdTension = rValue;
+        // COMPRESSION
+        else if (rVariable == DAMAGE_COMPRESSION)
+            mDamageCompression = rValue;
+        else if (rVariable == UNIAXIAL_STRESS_COMPRESSION)
+            mUniaxialStressCompression = rValue;
+        else if (rVariable == THRESHOLD_COMPRESSION)
+            mThresholdCompression = rValue;
+    }
+
+    void MasonryOrthotropicDamagePlaneStress2DLaw::InitializeMaterial(
+        const Properties& rProperties,
+        const GeometryType& rGeometry,
+        const Vector& rShapeFunctionsValues)
+    {
+        if (!InitializeDamageLaw) {
+            mThresholdTension = rProperties[YIELD_STRESS_TENSION];
+            mCurrentThresholdTension = mThresholdTension;
+
+            mThresholdCompression = rProperties[DAMAGE_ONSET_STRESS_COMPRESSION];
+            mCurrentThresholdCompression = mThresholdCompression;
+
+            mDamageTension = 0.0;
+            mDamageCompression = 0.0;
+            mUniaxialStressTension = 0.0;
+            mUniaxialStressCompression = 0.0;
+
+            this->ComputeCharacteristicLength(
+                rGeometry,
+                mInitialCharacteristicLength);
+
+            InitializeDamageLaw = true;
+        }
+    }
+
+    void MasonryOrthotropicDamagePlaneStress2DLaw::FinalizeMaterialResponseCauchy(
+        Parameters& rParameters)
+    {
+        // save converged values
+        mCurrentThresholdTension = mThresholdTension;
```

Comment:
the duplication of int vars can be avoided if they are updated only in the finalize sol step :)

Reply:
this probably a problem of the perturbation. I will have a look at the one which you have implemented recently to figure out if this can be applied, here.